src/mon/PaxosService.cc

```diff
@@ -393,15 +393,26 @@ void PaxosService::maybe_trim()
     return;
   }
 
-  const version_t trim_max = g_conf().get_val<version_t>("paxos_service_trim_max");
-  if (trim_max > 0 &&
-      to_remove > trim_max) {
-    dout(10) << __func__ << " trim_to " << trim_to << " would only trim " << to_remove
-	     << " > paxos_service_trim_max, limiting to " << trim_max
-	     << dendl;
-    trim_to = get_first_committed() + trim_max;
-    to_remove = trim_max;
-  }
+  to_remove = [to_remove, this] {
+    const version_t trim_max = g_conf().get_val<version_t>("paxos_service_trim_max");
+    if (trim_max == 0 || to_remove < trim_max) {
+      return to_remove;
+    }
+    if (to_remove < trim_max * 1.5) {
+      dout(10) << __func__ << " trim to " << get_trim_to() << " would only trim " << to_remove
+             << " > paxos_service_trim_max, limiting to " << trim_max
+             << dendl;
+      return trim_max;
+    }
+    const version_t new_trim_max = (trim_max + to_remove) / 2;
+    const uint64_t trim_max_multiplier = g_conf().get_val<uint64_t>("paxos_service_trim_max_multiplier");
+    if (trim_max_multiplier) {
+      return std::min(new_trim_max, trim_max * trim_max_multiplier);
+    } else {
+      return new_trim_max;
```

Comment:
shouldn't this be `return trim_max;` ?

Reply:
trim_max_multiplier is used to provide an upper bound for how much new_trim_max can grow. If it is not enabled we still go ahead and change the max, so we return new_trim_max and not trim_max.

Reply:
I suppose I am nitpicking, but the doc says "0 disables it". Isn't it more clear to say "1 disables it", since 1 will bring back the old behavior?

Reply:
So actually, nothing brings back the old behaviour. We stick with the old logic if to_remove is not much bigger than trim_max otherwise we increase trim_max. 
If trim_max_multiplier is 0, we just go ahead and increase trim_max and do not have any upper bound for this change. So with trim_max_multiplier as 0, we disable the upper bound check of new_trim_max with trim_max * trim_max_multiplier.

Reply:
Oh okay I see what you mean. If trim_max_multiplier is 1 the trim_max does not change.

Reply:
For the documentation, I thought of it as 0 disables any upper bound check for the new_trim_max and anything else enables it.

Reply:
Fare enough, I interpreted it as "0 disables the multiplication". If you want to make it crystal clear maybe the doc could be "(0 disables the upper bound check and will trim at the maximum rate)" ?